Track: bergshamra 0.4.x and wasmtime 36.x major upgrades

## Background

PR #670 bumped bergshamra from 0.3.1 to 0.4.0 and merged cleanly, but the v1.1.2 release workflow failed because Windows (x86_64-pc-windows-msvc) could not compile the new transitive dependency \`kryptering 0.1.0\`.

See PR #690 for the revert.

## Root cause

bergshamra 0.4.0 pulls in \`bergshamra-crypto 0.4.0\` -> \`kryptering 0.1.0\` -> \`cryptoki 0.7.0\`. The \`cryptoki\` crate defines \`types::Ulong\` based on the platform-specific \`CK_ULONG\` from PKCS11:

- Linux / macOS: \`CK_ULONG\` is \`unsigned long\` = 64-bit
- Windows: \`CK_ULONG\` is \`unsigned long\` = 32-bit (Windows keeps \`long\` as 32-bit even on 64-bit platforms)

\`kryptering\` calls \`(value: u64).into()\` to convert into \`Ulong\`, which works on Linux (where \`From<u64>\` is implemented) but fails on Windows (where only \`From<u32>\` is implemented).

\`\`\`
error[E0277]: the trait bound \`Ulong: From<u64>\` is not satisfied
  --> kryptering-0.1.0/src/pkcs11/mod.rs:147:30
\`\`\`

## Options

1. Wait for upstream fix in \`kryptering\` to use a platform-aware conversion (e.g., \`u64::try_into\` with fallback)
2. Convince \`bergshamra-crypto\` maintainers to gate PKCS11 behind an optional feature flag, since we only need SAML XML signature verification (not HSM support)
3. Fork \`kryptering\` or patch via \`[patch.crates-io]\` in Cargo.toml

## Action items

- [ ] File upstream issue on the kryptering repo
- [ ] File upstream issue on bergshamra-crypto requesting optional PKCS11 feature
- [ ] Re-evaluate bergshamra 0.4.x upgrade path once either fix lands

--- a/.cargo/audit.toml
+++ b/.cargo/audit.toml
@@ -13,25 +13,7 @@ ignore = [
     # Risk is low for our use case as we don't use MySQL or RSA authentication.
     "RUSTSEC-2023-0071",
 
-    # wasmtime 24.0.7 - Winch compiler backend advisories (2026-04-09 batch)
-    # These four advisories only affect users of the Winch compiler backend.
-    # We use the default Cranelift backend and never call Strategy::Winch.
-    # Tracked for future upgrade evaluation in issue #691.
-    "RUSTSEC-2026-0086", # Host data leakage with 64-bit tables and Winch (low, 2.3)
-    "RUSTSEC-2026-0089", # Host panic when Winch compiler executes table.fill (medium, 5.9)
-    "RUSTSEC-2026-0094", # Improperly masked return value from table.grow with Winch (medium, 6.1)
-    "RUSTSEC-2026-0095", # Winch sandbox-escaping memory access (critical, 9.0)
-
-    # wasmtime 24.0.7 - Pooling allocator data leakage (2026-04-09 batch)
-    # Only affects users of the pooling allocation strategy. We use the
-    # default on-demand allocator. Tracked in issue #691.
-    "RUSTSEC-2026-0088", # Data leakage between pooling allocator instances (low, 2.3)
-
-    # wasmtime 24.0.7 - aarch64 Cranelift sandbox escape (2026-04-09 batch)
-    # CRITICAL on aarch64 deployments. Fix requires upgrading to wasmtime
-    # >=36.0.7, which is a major version jump requiring code changes across
-    # the plugin system. WASM plugins can only be uploaded by authenticated
-    # admin users (not from untrusted sources), limiting exposure. Tracked
-    # for immediate follow-up in issue #691.
-    "RUSTSEC-2026-0096", # Miscompiled guest heap access, aarch64 Cranelift (critical, 9.0)
+    # wasmtime advisories from the 2026-04-09 batch (RUSTSEC-2026-0086,
+    # 0088, 0089, 0094, 0095, 0096) are resolved by the wasmtime 24 -> 36.0.7
+    # upgrade tracked in issue #691. No wasmtime ignores needed.
 ]

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -88,8 +88,8 @@ bytes = "1.5"
 clap = { version = "4.5", features = ["derive", "env"] }
 
 # WASM runtime
-wasmtime = { version = "24.0.7", features = ["component-model", "async"] }
-wasmtime-wasi = "24.0.7"
+wasmtime = { version = "36.0.7", features = ["component-model", "async"] }
+wasmtime-wasi = "36.0.7"
 
 # Git operations
 git2 = "0.20"

--- a/backend/src/services/wasm_bindings.rs
+++ b/backend/src/services/wasm_bindings.rs
@@ -16,7 +16,8 @@ pub mod v1 {
     wasmtime::component::bindgen!({
         world: "format-plugin",
         path: "src/wit/format-plugin.wit",
-        async: true,
+        imports: { default: async | trappable },
+        exports: { default: async },
     });
 }
 
@@ -25,7 +26,8 @@ pub mod v2 {
     wasmtime::component::bindgen!({
         world: "format-plugin-v2",
         path: "src/wit/format-plugin.wit",
-        async: true,
+        imports: { default: async | trappable },
+        exports: { default: async },
     });
 }
 

--- a/backend/src/services/wasm_runtime.rs
+++ b/backend/src/services/wasm_runtime.rs
@@ -11,7 +11,7 @@ use thiserror::Error;
 use tracing::{debug, error, info, warn};
 use wasmtime::component::{Component, Linker, ResourceTable};
 use wasmtime::{Config, Engine, ResourceLimiter, Store, StoreLimits, StoreLimitsBuilder};
-use wasmtime_wasi::{WasiCtx, WasiCtxBuilder, WasiView};
+use wasmtime_wasi::{WasiCtx, WasiCtxBuilder, WasiCtxView, WasiView};
 
 use crate::models::plugin::PluginResourceLimits;
 
@@ -101,12 +101,11 @@ impl PluginContext {
 }
 
 impl WasiView for PluginContext {
-    fn table(&mut self) -> &mut ResourceTable {
-        &mut self.resource_table
-    }
-
-    fn ctx(&mut self) -> &mut WasiCtx {
-        &mut self.wasi_ctx
+    fn ctx(&mut self) -> WasiCtxView<'_> {
+        WasiCtxView {
+            ctx: &mut self.wasi_ctx,
+            table: &mut self.resource_table,
+        }
     }
 }
 
@@ -122,9 +121,9 @@ impl ResourceLimiter for PluginContext {
 
     fn table_growing(
         &mut self,
-        current: u32,
-        desired: u32,
-        maximum: Option<u32>,
+        current: usize,
+        desired: usize,
+        maximum: Option<usize>,
     ) -> anyhow::Result<bool> {
         self.limits.table_growing(current, desired, maximum)
     }
@@ -216,7 +215,7 @@ impl WasmRuntime {
 
         // Add minimal WASI imports for basic I/O
         // We only expose wasi:io/streams for artifact data
-        wasmtime_wasi::add_to_linker_async(&mut linker)
+        wasmtime_wasi::p2::add_to_linker_async(&mut linker)
             .map_err(|e| WasmError::EngineError(format!("Failed to add WASI: {}", e)))?;
 
         info!("Compiled WASM component ({} bytes)", wasm_bytes.len());
@@ -1092,5 +1091,63 @@ mod tests {
     #[test]
     fn test_fuel_per_second_constant() {
         assert_eq!(FUEL_PER_SECOND, 100_000_000);
+    }
+
+    // -----------------------------------------------------------------------
+    // wasmtime 36.x migration - WasiView / WasiCtxView shape
+    // -----------------------------------------------------------------------
+
+    /// Exercises the new `WasiView::ctx(&mut self) -> WasiCtxView<'_>` shape
+    /// introduced in wasmtime-wasi 30.x (split out of the old dual-method
+    /// trait). Before the migration this used `table()` + `ctx()`.
+    #[test]
+    fn test_wasi_view_returns_ctx_view_with_both_refs() {
+        let limits = PluginResourceLimits::default();
+        let mut context = PluginContext::new("wv".to_string(), "wv-fmt".to_string(), &limits);
+
+        let view: WasiCtxView<'_> = context.ctx();
+        // Both refs must be populated - just de-reference them to prove
+        // they're live references into the same PluginContext.
+        let _ctx_ptr: *const WasiCtx = view.ctx;
+        let _table_ptr: *const wasmtime::component::ResourceTable = view.table;
+    }
+
+    /// Compiles a minimal valid component and creates a Store from it. This
+    /// exercises the full v36 code path: `Engine::new(&Config)`, component
+    /// parsing, `Linker::new`, `wasmtime_wasi::p2::add_to_linker_async`, and
+    /// `Store::new` with a `PluginContext` (which requires `WasiView`).
+    #[test]
+    fn test_compile_and_store_roundtrip_with_minimal_component() {
+        let runtime = WasmRuntime::new().unwrap();
+        // `(component)` is the minimal valid component in WAT text form.
+        let compiled = runtime
+            .compile(b"(component)")
+            .expect("minimal component should compile");
+
+        let limits = PluginResourceLimits::default();
+        let store = runtime
+            .create_store(&compiled, "rt-id", "rt-fmt", &limits)
+            .expect("store creation should succeed");
+
+        // If we got a Store back the WasiView + ResourceLimiter plumbing is
+        // wired up correctly for the new 36.x API.
+        assert_eq!(store.data().plugin_id, "rt-id");
+        assert_eq!(store.data().format_key, "rt-fmt");
+    }
+
+    /// Verifies the ResourceLimiter impl uses the 36.x `usize` signature for
+    /// `table_growing` (it was `u32` in 24.x). If this compiles, the shape
+    /// is correct; asserting the call returns Ok for a growth below the
+    /// default cap gives us confidence the delegation to StoreLimits still
+    /// works after the type change.
+    #[test]
+    fn test_resource_limiter_table_growing_usize_signature() {
+        let limits = PluginResourceLimits::default();
+        let mut context = PluginContext::new("rl".to_string(), "rl-fmt".to_string(), &limits);
+
+        let ok = context
+            .table_growing(0_usize, 16_usize, Some(10_000_usize))
+            .expect("table growth should be allowed under the default limit");
+        assert!(ok);
     }
 }

